test/functional/test_framework/address.py

```diff
@@ -41,7 +44,32 @@ def byte_to_base58(b, version):
         str = str[2:]
     return result
 
-# TODO: def base58_decode
+
+def base58_to_byte(s, version, verify_checksum=True):
+    if not s:
+        return b''
+    n = 0
+    for c in s:
+        n *= 58
+        assert c in chars
+        digit = chars.index(c)
+        n += digit
+    h = '%x' % n
+    if len(h) % 2:
+        h = '0' + h
+    res = n.to_bytes((n.bit_length() + 7) // 8, 'big')
+    pad = 0
+    for c in s:
+        if c == chars[0]:
+            pad += 1
+        else:
+            break
+    if (verify_checksum):
+        hash = byte_to_base58(res[1:-4], version)
+        assert_equal(s, hash)
+
+    return b'\x00' * pad + res[1:-4]
```

Comment:
In commit "tests: implement base58_decode" (31cfd81e1f374a722ab0190cd89814f22b8f91f8)

Might be nice return the version number as well as the data. Maybe:

```
res = b'\x00' * pad + res
if verify_checksum:
  assert_equal(hash256(res[:-4], res[-4:])
return res[1:-4], int(res[0])
```

I think this also fixes a padding bug. It looks like in `byte_to_base58` the padding precedes the version number and data, but the current return statement effectively adds the padding between the version number and the data.